lib/OpenQA/WebAPI/Controller/Admin/Needle.pm

```diff
@@ -74,69 +118,14 @@ sub ajax {
-                my $hash     = {
-                    id        => $n->id,
-                    directory => $n->directory->name,
-                    filename  => $filename,
-                };
-                my $needle_dir_path = $n->directory->path;
-                my $real_needle_dir_id;
-
-                if ($needle_dir_real_paths{$needle_dir_path}) {
-                    my $real_path = $needle_dir_real_paths{$needle_dir_path}->{real_path};
-                    $real_needle_dir_id = $needle_dir_real_paths{$needle_dir_path}->{real_path_id}
-                      if $real_path && $needle_dir_path ne $real_path;
-                }
-                else {
-                    my $real_path_id         = $n->directory->id;
-                    my $needle_dir_real_path = realpath($needle_dir_path);
-                    if ($needle_dir_real_path && $needle_dir_real_path ne $needle_dir_path) {
-                        my $real_needle_dir
-                          = $self->schema->resultset('NeedleDirs')->find({path => $needle_dir_real_path});
-                        if ($real_needle_dir) {
-                            $real_needle_dir_id = $real_needle_dir->id;
-                            $real_path_id       = $real_needle_dir_id;
-                        }
-                    }
-                    $needle_dir_real_paths{$needle_dir_path} = {
-                        real_path    => $needle_dir_real_path,
-                        real_path_id => $real_path_id,
-                    };
-                }
-                my $real_needle_info;
-                $real_needle_info
-                  = $schema->resultset('Needles')->find({dir_id => $real_needle_dir_id, filename => $filename})
-                  if $real_needle_dir_id;
-                $n = $real_needle_info if $real_needle_info;
-
-                $hash->{last_seen}  = $n->last_seen_time    || 'never';
-                $hash->{last_match} = $n->last_matched_time || 'never';
-
-                if (my $last_seen_module_id = $n->last_seen_module_id) {
-                    $hash->{last_seen_link} = $self->url_for(
-                        'admin_needle_module',
-                        module_id => $last_seen_module_id,
-                        needle_id => $n->id
-                    );
-                }
-                if (my $last_matched_module_id = $n->last_matched_module_id) {
-                    $hash->{last_match_link} = $self->url_for(
-                        'admin_needle_module',
-                        module_id => $last_matched_module_id,
-                        needle_id => $n->id
-                    );
-                }
-                push(@data, $hash);
-            }
-            return \@data;
+            [
+                map {
+                    $self->_prepare_data_table(
+                        $_, {},
+                        $self->schema->resultset('NeedleDirs'),
+                        $self->schema->resultset('Needles'))
```

Comment:
This still means get the result within the loop/map. My main point was to move the `$self->schema->resultset` calls outside the loop.

It looks also wrong to pass an empty hash-ref here. To have a shared hash for the paths you still need `%needle_dir_real_paths` outside the loop. And yes, it likely works this way as well but causes unnecessary lookups when I remember the algorithm correctly.

Reply:
Hm, I'm not sure about the resultset, might just be fine this way but hard to proof so I changed this. For the empty hash-ref, I think it is ok because if the hash would not be the very same for every iteration no values would be shared and tests would fail.